File: Data Structures & Algorithms/buy-and-sell-crypto/submission-2.go
func maxProfit(prices []int) int {
	if len(prices) == 1 { return 0 }

	profit := 0
	l, r := 0, 1
	for l < len(prices) && r < len(prices) {
		profit = max(profit, prices[r] - prices[l])
		if prices[r] > prices[l] {
			r++
		} else {
			l++
			r = max(r, l + 1)
		}
	}
	return profit
}

File: Data Structures & Algorithms/buy-and-sell-crypto-with-cooldown/submission-2.cs
public class Solution {
    public int MaxProfit(int[] prices) {
        if (prices.Length == 0) return 0;

        // Initial values for Day 0
        int holding = -prices[0]; // Profit after buying on day 0
        int sold = 0;             // Profit after selling (impossible on day 0)
        int free = 0;             // Profit while being free to buy

        for (int i = 1; i < prices.Length; i++) {
            int prevHolding = holding;
            int prevSold = sold;
            int prevFree = free;

            // 1. To be HOLDING: Either you held from yesterday, 
            //    OR you were FREE and you BUY today.
            holding = Math.Max(prevHolding, prevFree - prices[i]);

            // 2. To be in SOLD: You must have been HOLDING and you SELL today.
            sold = prevHolding + prices[i];

            // 3. To be FREE: Either you were already FREE, 
            //    OR you just finished your COOLDOWN (you were in SOLD yesterday).
            free = Math.Max(prevFree, prevSold);
        }

        // Your max profit will be either you just sold, or you're free.
        return Math.Max(sold, free);
    }
}
